server/routerlicious/packages/routerlicious-base/src/utils/sessionHelper.ts

```diff
@@ -134,20 +136,50 @@ async function updateExistingSession(
         isSessionActive: false,
     };
     try {
-        await documentsCollection.upsert(
+        const result = await documentsCollection.findAndUpdate(
             {
                 documentId,
+                "session.isSessionAlive": false,
             },
             {
+                createTime: document.createTime,
                 deli: updatedDeli ?? document.deli,
-                scribe: updatedScribe ?? document.scribe,
+                documentId: document.documentId,
                 session: updatedSession,
-            },
-            null);
+                scribe: updatedScribe ?? document.scribe,
+                tenantId: document.tenantId,
+                version: document.version,
+            });
         Lumberjack.info(
-            `The Session ${JSON.stringify(updatedSession)} was updated into the document collection`,
+            `The existing document: ${result} in the documents collection in updateExistingSession`,
             lumberjackProperties,
         );
+        // There is no document with isSessionAlive as false. It means this session has been discovered by
+        // another call, and there is a race condition with different clients writing truth into the cosmosdb
+        // from different clusters. Thus, let it call getSession for maximum three times.
+        if (!result.existing && count === 3) {
```

Comment:
I am wondering if we can simply use runWithRetry from services core here.

Reply:
runWithRetry is used to call API, this is just to call the internal method. So we can keep it as it is right now. Thanks.

Reply:
I actually like @pradeepvairamani's idea. Couldn't we make runWithRetry or requestWithRetry call this internal method? You could pass that method? That way, the logic would be less confusing and following the same standards we apply elsewhere in the code.

In fact, here it seems we have some kind of recursion. `updateExistingSession` would call `getSession`, which would call `updateExistingSession` again. Besides being a bit confusing, recursion can always present a risk of infinite loop (in case of unexpected bugs). Not saying it is the case here, but if possible, might be something we want to avoid.

Reply:
Since this method needs to use the recursive call, we are unable to use the runWithRetry call. The runWithRetry is used to call the method, again and again, not to call the method in recursion. 

Yes, this is intended to call getSession, which might call updateExistingSession again to a maximum of three trials. The loop would get out at the count of 3.

Reply:
The point I am trying to make, which I believe is also what Pradeep tried to say, is that if you look at the original implementation before this PR, I don't see a recursive call being made (please correct me if I missed something).

Instead of using this recursive logic here, couldn't you instead update `getSession` so that `getSession` would use  `runWithRetry` to call `updateExistingSession`? In other words, you would update line 259 to wrap `updateExistingSession` in `runWithRetry`. Is it not possible? Especially since `updateExistingSession` is already async.

If not possible, can you explain why this method needs the recursive call?

Reply:
Got it. I am going to introduce this recursion in the PR.
Nope for this logic, since we need to use the code logic in getSession in the condition of !result.existing && count < 3.

In most cases, we won't fall into the recursive condition. However, whenever the racing condition happens in the updateExistingSession 's cosmosDB write. 
Here is an example: 
1. In updateExistingSession method, Client A has already marked the session isSessionAlive as true. 
2. In updateExistingSession method, Client B would read the result.existing as false since we have the isSessionAlive check flag. It calls the getSession to read the session. 
          2.1.In the getSession method: if the isSessionAlive is true, return the session URLs. Otherwise, got to updateExistingSession again, and try to update and return the correct URLs. 

We do the recursion at #2 for three times to run for more trials. But mostly, at 2.1, isSessionAlive has already been updated as true, and would return the result at that level.

Reply:
I directly removed this logic, and make the call to the database again. :)